File: app/tests/mock/engine/random_mover.cpp
#include <random>

#include "../../../third_party/chess.hpp"

using namespace chess;

std::mt19937 gen;

void initialize_rng(int argc, char *argv[]) {
    if (argc > 1) {
        std::cout << "Using provided seed: " << argv[1] << std::endl;
        unsigned int seed = std::stoi(argv[1]);
        gen.seed(seed);
    } else {
        std::random_device rd;
        auto seed = rd();
        gen.seed(seed);
        std::cout << "Using random seed: " << seed << std::endl;
    }
}

int random_number(int min, int max) {
    std::uniform_int_distribution<> dis(min, max);
    return dis(gen);
}

Move random_move(const Board &board) {
    auto moves = Movelist();
    movegen::legalmoves(moves, board);

    return moves[random_number(0, moves.size() - 1)];
}

void uci_line(Board &board, const std::string &line) {
    auto tokens = utils::splitString(line, ' ');

    if (tokens.empty())
        return;
    else if (tokens[0] == "position") {
        // parse position
        if (tokens[1] == "startpos") {
            board = Board();
            if (tokens.size() > 2 && tokens[2] == "moves") {
                for (int i = 3; i < tokens.size(); i++) {
                    board.makeMove(uci::uciToMove(board, std::string(tokens[i])));
                }
            }
        } else if (tokens[1] == "fen") {
            std::string fen = "";
            // parse until "moves"
            int i = 2;
            for (; i < tokens.size(); i++) {
                if (tokens[i] == "moves") {
                    i++;
                    break;
                }
                fen += std::string(tokens[i]) + " ";
            }

            board.setFen(fen);

            for (; i < tokens.size(); i++) {
                board.makeMove(uci::uciToMove(board, std::string(tokens[i])));
            }
        }
    } else if (tokens[0] == "uci") {
        std::cout << "id name random_move" << std::endl;
        std::cout << "id author fastchess" << std::endl;
        std::cout << "option name Hash type spin default 16 min 1 max 33554432" << std::endl;
        std::cout << "uciok" << std::endl;
    } else if (tokens[0] == "isready") {
        std::cout << "readyok" << std::endl;
    } else if (tokens[0] == "ucinewgame") {
        board = Board();
    } else if (tokens[0] == "go") {
        const auto move = random_move(board);
        std::cout << "info depth 1 pv " << uci::moveToUci(move) << " score 0 " << std::endl;
        std::cout << "bestmove " << uci::moveToUci(move) << std::endl;
    }
}

void uci_loop() {
    std::string input;
    std::cin >> std::ws;

    Board board = Board();

    while (true) {
        if (!std::getline(std::cin, input) && std::cin.eof()) {
            input = "quit";
        }

        if (input == "quit") {
            return;
        } else {
            uci_line(board, input);
        }
    }
}

int main(int argc, char *argv[]) {
    initialize_rng(argc, argv);
    uci_loop();
}

File: app/tests/mock/engine/random_mover.rs
use chess_library_rs::board::Board;
use chess_library_rs::movegen;
use chess_library_rs::movelist::Movelist;
use chess_library_rs::uci;

use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use std::io::{self, BufRead, Write};

struct RandomMover {
    rng: StdRng,
    board: Board,
}

impl RandomMover {
    fn new(seed: Option<u64>) -> Self {
        let rng = match seed {
            Some(s) => {
                eprintln!("Using provided seed: {}", s);
                StdRng::seed_from_u64(s)
            }
            None => {
                let s = rand::random::<u64>();
                eprintln!("Using random seed: {}", s);
                StdRng::seed_from_u64(s)
            }
        };

        RandomMover {
            rng,
            board: Board::new(),
        }
    }

    fn random_move(&mut self) -> String {
        let mut moves = Movelist::new();
        movegen::legalmoves(
            &mut moves,
            &self.board,
            movegen::MoveGenType::All,
            movegen::PieceGenType::ALL,
        );

        if moves.is_empty() {
            return "0000".to_string();
        }

        let idx = self.rng.gen_range(0..moves.size());
        let mv = moves[idx];
        uci::move_to_uci(mv, self.board.chess960())
    }

    fn handle_position(&mut self, line: &str) {
        let tokens: Vec<&str> = line.split_whitespace().collect();
        if tokens.len() < 2 {
            return;
        }

        if tokens[1] == "startpos" {
            self.board = Board::new();
            if tokens.len() > 2 && tokens[2] == "moves" {
                for i in 3..tokens.len() {
                    let mv = uci::uci_to_move(&self.board, tokens[i]);
                    if mv.raw() != 0 {
                        self.board.make_move(mv, false);
                    }
                }
            }
        } else if tokens[1] == "fen" {
            let mut fen_parts = Vec::new();
            let mut i = 2;
            while i < tokens.len() && tokens[i] != "moves" {
                fen_parts.push(tokens[i]);
                i += 1;
            }
            let fen = fen_parts.join(" ");
            self.board.set_fen(&fen);

            if i < tokens.len() && tokens[i] == "moves" {
                i += 1;
                while i < tokens.len() {
                    let mv = uci::uci_to_move(&self.board, tokens[i]);
                    if mv.raw() != 0 {
                        self.board.make_move(mv, false);
                    }
                    i += 1;
                }
            }
        }
    }

    fn handle_uci(&self) {
        println!("id name random_move");
        println!("id author fastchess");
        println!("option name Hash type spin default 16 min 1 max 33554432");
        println!("uciok");
    }

    fn handle_isready(&self) {
        println!("readyok");
    }

    fn handle_ucinewgame(&mut self) {
        self.board = Board::new();
    }

    fn handle_go(&mut self) {
        let mv = self.random_move();
        println!("info depth 1 pv {} score cp 0", mv);
        println!("bestmove {}", mv);
    }

    fn run(&mut self) {
        let stdin = io::stdin();
        let stdout = io::stdout();
        let mut stdout_lock = stdout.lock();

        for line in stdin.lock().lines() {
            let line = match line {
                Ok(l) => l,
                Err(_) => break,
            };

            let line = line.trim();

            if line == "quit" {
                break;
            } else if line == "uci" {
                self.handle_uci();
            } else if line == "isready" {
                self.handle_isready();
            } else if line == "ucinewgame" {
                self.handle_ucinewgame();
            } else if line.starts_with("position") {
                self.handle_position(line);
            } else if line.starts_with("go") {
                self.handle_go();
            }

            stdout_lock.flush().unwrap();
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    let seed = args.get(1).and_then(|s| s.parse::<u64>().ok());
    let mut mover = RandomMover::new(seed);
    mover.run();
}
